Log Chain ID in Events When Applying ChainPolicy in FleetAutoscaler

0xaravindh opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://github.com/googleforgames/agones/blob/main/CONTRIBUTING.md and developer guide https://github.com/googleforgames/agones/blob/main/build/README.md
2. Please label this pull request according to what type of issue you are addressing.
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/googleforgames/agones/blob/main/build/README.md#testing-and-building
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug

/kind cleanup

> /kind documentation
> /kind feature
> /kind hotfix
> /kind release

**What this PR does / Why we need it**:

Added logging of Chain ID to the event message when a ChainPolicy is applied in the FleetAutoscaler controller.


**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Closes #<issue number>`, or `Closes (paste link of issue)`.
-->
Closes #3953

**Special notes for your reviewer**:

vicentefb commented:
@indexjoseph fyi

0xaravindh commented:
Below is an example of the relevant event output:

```
karavindh@karavindh:~$ k describe fleetautoscaler chain-fleet-autoscaler 
Name:         chain-fleet-autoscaler
Namespace:    default
Labels:       <none>
Annotations:  <none>
API Version:  autoscaling.agones.dev/v1
Kind:         FleetAutoscaler
Metadata:
  Creation Timestamp:  2025-03-21T18:27:33Z
  Generation:          1
  Resource Version:    1742581683798655009
  UID:                 668b9aa1-3905-40c1-989d-28c256ca7b16
Spec:
  Fleet Name:  fleet-example
  Policy:
    Chain:
      Id:  weekday
      Schedule:
        Active Period:
          Duration:    5h
          Start Cron:  0 1 * * 0
          Timezone:    America/New_York
        Between:
          End:    2100-02-23T21:04:04Z
          Start:  2100-02-20T21:04:04Z
        Policy:
          Buffer:
            Buffer Size:   50
            Max Replicas:  2000
            Min Replicas:  100
          Type:            Buffer
      Type:                Schedule
      Id:                  weekend
      Schedule:
        Active Period:
          Duration:    7h
          Start Cron:  0 1 * * 0
          Timezone:    America/New_York
        Between:
          End:    2100-02-26T21:04:05Z
          Start:  2100-02-24T21:04:05Z
        Policy:
          Counter:
            Buffer Size:   10
            Key:           rooms
            Max Capacity:  1000
            Min Capacity:  500
          Type:            Counter
      Type:                Schedule
      Buffer:
        Buffer Size:   5
        Max Replicas:  2000
        Min Replicas:  100
      Id:              default
      Type:            Buffer
    Type:              Chain
  Sync:
    Fixed Interval:
      Seconds:  30
    Type:       FixedInterval
Status:
  Able To Scale:     true
  Current Replicas:  100
  Desired Replicas:  100
  Last Scale Time:   2025-03-21T18:27:33Z
  Scaling Limited:   true
Events:
  Type     Reason              Age                  From                        Message
  ----     ------              ----                 ----                        -------
  Normal   AutoScalingFleet    10m                  fleetautoscaler-controller  Scaling fleet fleet-example from 2 to 100
  Warning  ScalingLimited      9m43s (x2 over 10m)  fleetautoscaler-controller  Scaling fleet fleet-example was limited to minimum size of 100
  Normal   ChainPolicyApplied  13s (x21 over 10m)   fleetautoscaler-controller  FleetAutoscaler 'chain-fleet-autoscaler' applied ChainPolicy | ID: default | Type: Buffer
karavindh@karavindh:~$ 
```

igooch commented:
/gcbrun

indexjoseph commented:
I'm also curious if this bug can be consolidated with #3955.

0xaravindh commented:
> #3955



> I think this all looks good except for one thing - it seems like the current functionality we're logging on every sync cycle. I'm not sure if it matters that much but I think we should only log when there's a change within the chain. For instance, we'd log when the `FleetAutoscaler` switches from applying the `default` policy to applying the `weekend` policy.

https://github.com/0xaravindh/agones/blob/09ca94005be43ae0159631177d0f51f086418530/pkg/fleetautoscalers/fleetautoscalers.go#L465-L470

You want to stop logging the default state. right?

indexjoseph commented:
> > #3955
> 
> > I think this all looks good except for one thing - it seems like the current functionality we're logging on every sync cycle. I'm not sure if it matters that much but I think we should only log when there's a change within the chain. For instance, we'd log when the `FleetAutoscaler` switches from applying the `default` policy to applying the `weekend` policy.
> 
> https://github.com/0xaravindh/agones/blob/09ca94005be43ae0159631177d0f51f086418530/pkg/fleetautoscalers/fleetautoscalers.go#L465-L470
> 
> You want to stop logging the default state. right?

Not necessarily, I'm pointing out that the current implementation logs the applied policy and fleet size on every synchronization cycle of the FleetAutoscaler. This results in repetitive event entries, even when there's no actual change in the applied policy or the fleet's size.

My suggestion is to modify the logging behavior so that it only occurs when the FleetAutoscaler switches from applying one policy to another (e.g., from 'default' to 'weekend'). Although I'm unsure if this is possible without adding logic to hold state, there's no reliable way to compare the previous policy within a chain against the current policy and determine whether a new log entry is needed, it's up to you if you think it's necessary.

0xaravindh commented:
> > > #3955
> > 
> > 
> > > I think this all looks good except for one thing - it seems like the current functionality we're logging on every sync cycle. I'm not sure if it matters that much but I think we should only log when there's a change within the chain. For instance, we'd log when the `FleetAutoscaler` switches from applying the `default` policy to applying the `weekend` policy.
> > 
> > 
> > https://github.com/0xaravindh/agones/blob/09ca94005be43ae0159631177d0f51f086418530/pkg/fleetautoscalers/fleetautoscalers.go#L465-L470
> > You want to stop logging the default state. right?
> 
> Not necessarily, I'm pointing out that the current implementation logs the applied policy and fleet size on every synchronization cycle of the FleetAutoscaler. This results in repetitive event entries, even when there's no actual change in the applied policy or the fleet's size.
> 
> My suggestion is to modify the logging behavior so that it only occurs when the FleetAutoscaler switches from applying one policy to another (e.g., from 'default' to 'weekend'). Although I'm unsure if this is possible without adding logic to hold state, as of right now there's no way to compare the previous policy within a chain against the current policy and determine whether a new log entry is needed, it's up to you if you think it's necessary.

@indexjoseph I added a `LastAppliedPolicy` field to track the previously applied policy. Now, logs will only show when the policy actually changes (e.g., from default to weekend). This should prevent the repeated event entries you pointed out. Let me know if that works!

```
karavindh@karavindh:~$ kubectl describe fleetautoscaler chain-fleet-autoscaler
Name:         chain-fleet-autoscaler
Namespace:    default
Labels:       <none>
Annotations:  <none>
API Version:  autoscaling.agones.dev/v1
Kind:         FleetAutoscaler
Metadata:
  Creation Timestamp:  2025-04-16T07:11:28Z
  Generation:          1
  Resource Version:    1744787518525023009
  UID:                 8a835d61-5578-4a84-85e1-861570bd61ab
Spec:
  Fleet Name:  fleet-example
  Policy:
    Chain:
      Id:  weekday
      Schedule:
        Active Period:
          Duration:    5h
          Start Cron:  0 0 * * *
          Timezone:    America/Los_Angeles
        Between:
          End:    2025-07-01T07:00:00Z
          Start:  2025-04-01T07:00:00Z
        Policy:
          Buffer:
            Buffer Size:   50
            Max Replicas:  120
            Min Replicas:  30
          Type:            Buffer
      Type:                Schedule
      Id:                  weekend
      Schedule:
        Active Period:
          Duration:    5h
          Start Cron:  0 0 * * *
          Timezone:    America/Los_Angeles
        Between:
          End:    2025-07-01T07:00:00Z
          Start:  2025-04-01T07:00:00Z
        Policy:
          Counter:
            Buffer Size:   10
            Key:           rooms
            Max Capacity:  120
            Min Capacity:  30
          Type:            Counter
      Type:                Schedule
      Buffer:
        Buffer Size:   5
        Max Replicas:  100
        Min Replicas:  20
      Id:              default
      Type:            Buffer
    Type:              Chain
  Sync:
    Fixed Interval:
      Seconds:  30
    Type:       FixedInterval
Status:
  Able To Scale:        true
  Current Replicas:     50
  Desired Replicas:     50
  Last Applied Policy:  Chain:weekday:Schedule
  Last Scale Time:      2025-04-16T07:11:28Z
  Scaling Limited:      false
Events:
  Type    Reason            Age   From                        Message
  ----    ------            ----  ----                        -------
  Normal  ChainPolicy       20m   fleetautoscaler-controller  FleetAutoscaler 'chain-fleet-autoscaler' successfully applied ChainPolicy | ID: weekday | Type: Schedule
  Normal  AutoScalingFleet  20m   fleetautoscaler-controller  Scaling fleet fleet-example from 2 to 50
karavindh@karavindh:~$ 
```

0xaravindh commented:
/gcbrun

indexjoseph commented:
> @indexjoseph I added a `LastAppliedPolicy` field to track the previously applied policy. Now, logs will only show when the policy actually changes (e.g., from default to weekend). This should prevent the repeated event entries you pointed out. Let me know if that works!

Yep, LGTM!

vicentefb commented on pkg/apis/autoscaling/v1/fleetautoscaler.go:
```diff
@@ -301,6 +301,11 @@ type FleetAutoscalerStatus struct {
+
+	// LastAppliedPolicy is the ID of the last applied policy in the ChainPolicy.
+	// Used to track policy transitions for logging purposes.
+	// +optional
```
why is this optional ?

0xaravindh replied:
removed!

vicentefb commented on pkg/fleetautoscalers/fleetautoscalers.go:
```diff
@@ -253,9 +263,36 @@ func applyBufferPolicy(b *autoscalingv1.BufferPolicy, f *agonesv1.Fleet) (int32,
 }
 
+// New function to call applyCounterOrListPolicy
+func applyCounterOrListPolicyWrapper(c *autoscalingv1.CounterPolicy, l *autoscalingv1.ListPolicy,
```
The logging in `applyCounterOrListPolicyWrapper` happens only if applyCounterOrListPolicy returns nil for the error, right? 

This means successful applications of the policy are logged, but failures within `applyCounterOrListPolicy` are not logged with the same "Fleet Autoscaler operation completed" message. You might want to log regardless of the error status to ensure all attempts are recorded (?)

0xaravindh replied:
I haven't added any error loggin in the fleetautoscalers.go file. So, if something goes wrong, the default error handling will take care of it.

https://github.com/0xaravindh/agones/blob/490d764eaa0a78041209369097c6409bd48661fb/pkg/fleetautoscalers/controller.go#L343-L344

vicentefb replied:
Yeah, maybe we can create an Issue to enhance logging so that we log regardless of the error status to ensure all attempts are recorded. Not sure if this could clog the log.

vicentefb commented on pkg/fleetautoscalers/fleetautoscalers.go:
```diff
@@ -214,13 +215,22 @@ func applyWebhookPolicy(w *autoscalingv1.WebhookPolicy, f *agonesv1.Fleet) (repl
 }
 
-func applyBufferPolicy(b *autoscalingv1.BufferPolicy, f *agonesv1.Fleet) (int32, bool, error) {
+func applyBufferPolicy(b *autoscalingv1.BufferPolicy, f *agonesv1.Fleet, fasLog *FasLogger) (int32, bool, error) {
```
Just confirming and double checking, could there be a potential integer overflow ? This is out of scope for this PR but just wondering.

0xaravindh replied:
We can’t pass too many parameters easily in a function, so I put them into a separate struct. what are the cons here? Could you explain a little bit more?

vicentefb replied:
Nevermind, I was imagining a very extreme scenario. For example, if AllocatedReplicas = 2,000,000,000 and BufferSize (as int32) = 200,000,000, the sum 2,200,000,000 would overflow int32.

vicentefb commented on pkg/fleetautoscalers/fleetautoscalers.go:
```diff
@@ -404,31 +444,50 @@ func applyChainPolicy(c autoscalingv1.ChainPolicy, f *agonesv1.Fleet, gameServer
+	var chainEntry autoscalingv1.FleetAutoscalerPolicyType
 
 	// Loop over all entries in the chain
 	for _, entry := range c {
```
The applyChainPolicy iterates through the policies in the chain. If an error occurs for a specific policy, the loop continues. Only if all policies in the chain fail will the function return an error. The logging and event emission in `applyChainPolicy` happen only if an error occurs or if there's a change in the applied policy. Is this interpretation correct @indexjoseph I might have forgotten.

vicentefb reviewed (commented):
I left some questions

igooch commented on install/helm/agones/templates/crds/fleetautoscaler.yaml:
```diff
@@ -93,6 +93,8 @@ spec:
                 scalingLimited:
                   type: boolean
+                lastAppliedPolicy:
+                  type: string
```
Per the CRD policy, this field either needs to be optional, or have a default value. In this case it looks like the controller needs to have this field, so there should be a default value. https://github.com/googleforgames/agones/edit/main/install/helm/agones/README.md has the details.

```suggestion
                  type: string
                  default: ""
```

0xaravindh replied:
Updated now!

igooch commented on pkg/client/applyconfiguration/autoscaling/v1/fleetautoscalerstatus.go:
```diff
@@ -77,3 +79,11 @@ func (b *FleetAutoscalerStatusApplyConfiguration) WithScalingLimited(value bool)
 	return b
 }
+
+// WithLastAppliedPolicy sets the LastAppliedPolicy field in the declarative configuration to the given value
+// and returns the receiver, so that objects can be built by chaining "With" function invocations.
+// If called multiple times, the LastAppliedPolicy field is set to the value of the last call.
```
The first sentence of this comment is confusing. Could you reword?

0xaravindh replied:
we can't edit this file. please correct me if i'm wrong

https://github.com/googleforgames/agones/blob/533ff3fc0ef29cd8424466012796bab45759ed19/pkg/client/applyconfiguration/autoscaling/v1/fleetautoscalerstatus.go#L15-L17

igooch reviewed (changes requested):
Is there a unit or e2e test that checks that the last applied policy works as expected? And throws handles errors as expected?

0xaravindh commented:
> Is there a unit or e2e test that checks that the last applied policy works as expected? And throws handles errors as expected?

Yes, I added unit and e2e tests to verify LastAppliedPolicy and error handling.

0xaravindh commented:
/gcbrun

vicentefb reviewed (approved):
lgtm, i believe Ivy requested changes

0xaravindh commented on test/e2e/fleetautoscaler_test.go:
```diff
@@ -1301,6 +1301,20 @@ func TestCounterAutoscalerAllocatedMultipleNamespaces(t *testing.T) {
 			})
 
+			// Wait for LastAppliedPolicy to be set to CounterPolicyType for fasA
+			framework.WaitForFleetAutoScalerCondition(t, fasA, func(log *logrus.Entry, fas *autoscalingv1.FleetAutoscaler) bool {
```
I added couple of testcase for `TestChainAutoscaler` and `TestAutoscalerWebhook`

igooch reviewed (approved):
LGTM
